File: WorldFeed/Server/src/Services/Post/WorldFeed.Post.API/Client/Clients/TimelinesClient.cs
namespace WorldFeed.Post.Client.Clients
{
    using System.Linq;
    using System.Threading.Tasks;

    using WorldFeed.Common.Client.Validators;
    using WorldFeed.Common.Iterators;
    using WorldFeed.Common.Public.Iterators;
    using WorldFeed.Common.Public.Models.Interfaces;
    using WorldFeed.Common.Public.Models.Interfaces.DTO;
    using WorldFeed.Post.Application.Parameters.TimelineClient;

    public class TimelinesClient : ITimelinesClient
    {
        private readonly ITwitterClient client;
        private readonly ITimelinesRequester timelinesRequester;

        public TimelinesClient(ITwitterClient client)
        {
            this.client = client;
            this.timelinesRequester = this.client.Raw.Timelines;
        }

        public ITimelineClientParametersValidator ParametersValidator => this.client.ParametersValidator;
        public Task<ITweet[]> GetHomeTimelineAsync()
        {
            return GetHomeTimelineAsync(new GetHomeTimelineParameters());
        }

        public async Task<ITweet[]> GetHomeTimelineAsync(IGetHomeTimelineParameters parameters)
        {
            var iterator = GetHomeTimelineIterator(parameters);
            return (await iterator.NextPageAsync().ConfigureAwait(false)).ToArray();
        }

        public ITwitterIterator<ITweet, long?> GetHomeTimelineIterator()
        {
            return GetHomeTimelineIterator(new GetHomeTimelineParameters());
        }

        public ITwitterIterator<ITweet, long?> GetHomeTimelineIterator(IGetHomeTimelineParameters parameters)
        {
            var pageIterator = this.timelinesRequester.GetHomeTimelineIterator(parameters);
            return new TwitterIteratorProxy<ITwitterResult<ITweetDTO[]>, ITweet, long?>(pageIterator,
                twitterResult => this.client.Factories.CreateTweets(twitterResult?.Model));
        }

        public Task<ITweet[]> GetUserTimelineAsync(long userId)
        {
            return GetUserTimelineAsync(new GetUserTimelineParameters(userId));
        }

        public Task<ITweet[]> GetUserTimelineAsync(string username)
        {
            return GetUserTimelineAsync(new GetUserTimelineParameters(username));
        }

        public Task<ITweet[]> GetUserTimelineAsync(IUserIdentifier user)
        {
            return GetUserTimelineAsync(new GetUserTimelineParameters(user));
        }

        public async Task<ITweet[]> GetUserTimelineAsync(IGetUserTimelineParameters parameters)
        {
            var iterator = GetUserTimelineIterator(parameters);
            return (await iterator.NextPageAsync().ConfigureAwait(false)).ToArray();
        }

        public ITwitterIterator<ITweet, long?> GetUserTimelineIterator(long userId)
        {
            return GetUserTimelineIterator(new GetUserTimelineParameters(userId));
        }

        public ITwitterIterator<ITweet, long?> GetUserTimelineIterator(string username)
        {
            return GetUserTimelineIterator(new GetUserTimelineParameters(username));
        }

        public ITwitterIterator<ITweet, long?> GetUserTimelineIterator(IUserIdentifier user)
        {
            return GetUserTimelineIterator(new GetUserTimelineParameters(user));
        }

        public ITwitterIterator<ITweet, long?> GetUserTimelineIterator(IGetUserTimelineParameters parameters)
        {
            var pageIterator = this.timelinesRequester.GetUserTimelineIterator(parameters);

            return new TwitterIteratorProxy<ITwitterResult<ITweetDTO[]>, ITweet, long?>(pageIterator,
                twitterResult => this.client.Factories.CreateTweets(twitterResult?.Model));
        }

        public Task<ITweet[]> GetMentionsTimelineAsync()
        {
            return GetMentionsTimelineAsync(new GetMentionsTimelineParameters());
        }

        public async Task<ITweet[]> GetMentionsTimelineAsync(IGetMentionsTimelineParameters parameters)
        {
            var iterator = GetMentionsTimelineIterator(parameters);
            return (await iterator.NextPageAsync().ConfigureAwait(false)).ToArray();
        }

        public ITwitterIterator<ITweet, long?> GetMentionsTimelineIterator()
        {
            return GetMentionsTimelineIterator(new GetMentionsTimelineParameters());
        }

        public ITwitterIterator<ITweet, long?> GetMentionsTimelineIterator(IGetMentionsTimelineParameters parameters)
        {
            var pageIterator = this.timelinesRequester.GetMentionsTimelineIterator(parameters);
            return new TwitterIteratorProxy<ITwitterResult<ITweetDTO[]>, ITweet, long?>(pageIterator,
                twitterResult => this.client.Factories.CreateTweets(twitterResult?.Model));
        }

        public Task<ITweet[]> GetRetweetsOfMeTimelineAsync()
        {
            return GetRetweetsOfMeTimelineAsync(new GetRetweetsOfMeTimelineParameters());
        }

        public async Task<ITweet[]> GetRetweetsOfMeTimelineAsync(IGetRetweetsOfMeTimelineParameters parameters)
        {
            var iterator = GetRetweetsOfMeTimelineIterator(parameters);
            var firstResults = await iterator.NextPageAsync().ConfigureAwait(false);
            return firstResults?.ToArray();
        }

        public ITwitterIterator<ITweet, long?> GetRetweetsOfMeTimelineIterator()
        {
            return GetRetweetsOfMeTimelineIterator(new GetRetweetsOfMeTimelineParameters());
        }

        public ITwitterIterator<ITweet, long?> GetRetweetsOfMeTimelineIterator(IGetRetweetsOfMeTimelineParameters parameters)
        {
            var pageIterator = this.timelinesRequester.GetRetweetsOfMeTimelineIterator(parameters);

            return new TwitterIteratorProxy<ITwitterResult<ITweetDTO[]>, ITweet, long?>(pageIterator,
                twitterResult => this.client.Factories.CreateTweets(twitterResult?.Model));
        }
    }
}

File: Sharebook/Client/src/app/sharebook/Client/Clients/TimelinesClient.ts
import {Inject} from "@angular/core";

import {ITwitterResult} from "../../../core/Core/Web/TwitterResult";
import {IUserIdentifier} from "../../../core/Public/Models/Interfaces/IUserIdentifier";
import {ITimelinesClient} from "../../../core/Public/Client/Clients/ITimelinesClient";
import {ITwitterIterator} from "../../../core/Public/Iterators/ITwitterIterator";
import {ITwitterClient, ITwitterClientToken} from "../../../core/Public/ITwitterClient";
import {ITimelinesRequester} from "../../../core/Public/Client/Requesters/ITimelinesRequester";
import {ITimelineClientParametersValidator} from "../../../core/Core/Client/Validators/TimelineClientParametersValidator";
import {ITweet} from "../../../core/Public/Models/Interfaces/ITweet";
import {
  GetHomeTimelineParameters,
  IGetHomeTimelineParameters
} from "../../../core/Public/Parameters/TimelineClient/GetHomeTimelineParameters";
import {ITweetDTO} from "../../../core/Public/Models/Interfaces/DTO/ITweetDTO";
import {TwitterIteratorProxy} from "../../../core/Core/Iterators/TwitterIteratorProxy";
import {
  GetUserTimelineParameters,
  IGetUserTimelineParameters
} from "../../../core/Public/Parameters/TimelineClient/GetUserTimelineParameters";
import {
  GetMentionsTimelineParameters,
  IGetMentionsTimelineParameters
} from "../../../core/Public/Parameters/TimelineClient/GetMentionsTimelineParameters";
import {
  GetRetweetsOfMeTimelineParameters,
  IGetRetweetsOfMeTimelineParameters
} from "../../../core/Public/Parameters/TimelineClient/GetRetweetsOfMeTimelineParameters";
import Type from "typescript-dotnet-commonjs/System/Types";
import {TwitterClient} from "../../TwitterClient";

export class TimelinesClient implements ITimelinesClient {
  private readonly _client: ITwitterClient;
  private readonly _timelinesRequester: ITimelinesRequester;

  constructor(/*@Inject(ITwitterClientToken) *//*client: TwitterClient*/) {
    // this._client = client;
    // this._timelinesRequester = this._client.raw.timelines;
  }

  get parametersValidator(): ITimelineClientParametersValidator {
    return this._client.parametersValidator;
  }

  public async getHomeTimelineAsync(parameters?: IGetHomeTimelineParameters): Promise<ITweet[]> {
    let parametersCurrent: IGetHomeTimelineParameters;
    if (parameters) {
      parametersCurrent = parameters;
    } else {
      parametersCurrent = new GetHomeTimelineParameters();
    }

    let iterator = this.getHomeTimelineIterator(parametersCurrent);
    return [...(await iterator.nextPageAsync())];            // .ConfigureAwait(false)).ToArray();
  }

  public getHomeTimelineIterator(parameters?: IGetHomeTimelineParameters): ITwitterIterator<ITweet, number> {    // long?
    let parametersCurrent: IGetHomeTimelineParameters;
    if (parameters) {
      parametersCurrent = parameters;
    } else {
      parametersCurrent = new GetHomeTimelineParameters();
    }

    let pageIterator = this._timelinesRequester.getHomeTimelineIterator(parametersCurrent);
    return new TwitterIteratorProxy<ITwitterResult<ITweetDTO[]>, ITweet, number>(pageIterator, twitterResult =>
      this._client.factories.createTweets(twitterResult?.model));
  }

  public async getUserTimelineAsync(userIdOrUsernameOrUserIdentifierOrParameters: number | string | IUserIdentifier | IGetUserTimelineParameters): Promise<ITweet[]> {
    let parameters: IGetUserTimelineParameters;
    if (Type.isNumber(userIdOrUsernameOrUserIdentifierOrParameters) || Type.isString(userIdOrUsernameOrUserIdentifierOrParameters)
      || this.isIUserIdentifier(userIdOrUsernameOrUserIdentifierOrParameters)) {
      parameters = new GetUserTimelineParameters(userIdOrUsernameOrUserIdentifierOrParameters);
    } else {
      parameters = userIdOrUsernameOrUserIdentifierOrParameters;
    }

    let iterator = this.getUserTimelineIterator(parameters);
    return [...(await iterator.nextPageAsync())]; // .ConfigureAwait(false)).ToArray();
  }

  public getUserTimelineIterator(userIdOrUsernameOrUserIdentifierOrParameters: number | string | IUserIdentifier | IGetUserTimelineParameters): ITwitterIterator<ITweet, number> {   // long?
    let parameters: IGetUserTimelineParameters;
    if (Type.isNumber(userIdOrUsernameOrUserIdentifierOrParameters) || Type.isString(userIdOrUsernameOrUserIdentifierOrParameters)
      || this.isIUserIdentifier(userIdOrUsernameOrUserIdentifierOrParameters)) {
      parameters = new GetUserTimelineParameters(userIdOrUsernameOrUserIdentifierOrParameters);
    } else {
      parameters = userIdOrUsernameOrUserIdentifierOrParameters;
    }

    let pageIterator = this._timelinesRequester.getUserTimelineIterator(parameters);

    return new TwitterIteratorProxy<ITwitterResult<ITweetDTO[]>, ITweet, number>(pageIterator, twitterResult =>
      this._client.factories.createTweets(twitterResult?.model));
  }

  public async getMentionsTimelineAsync(parameters?: IGetMentionsTimelineParameters): Promise<ITweet[]> {
    let parametersCurrent: IGetMentionsTimelineParameters;
    if (parameters) {
      parametersCurrent = parameters;
    } else {
      parametersCurrent = new GetMentionsTimelineParameters();
    }

    let iterator = this.getMentionsTimelineIterator(parametersCurrent);
    return [...(await iterator.nextPageAsync())]; // .ConfigureAwait(false)).ToArray();
  }

  public getMentionsTimelineIterator(parameters?: IGetMentionsTimelineParameters): ITwitterIterator<ITweet, number> {    // long?
    let parametersCurrent: IGetMentionsTimelineParameters;
    if (parameters) {
      parametersCurrent = parameters;
    } else {
      parametersCurrent = new GetMentionsTimelineParameters();
    }

    let pageIterator = this._timelinesRequester.getMentionsTimelineIterator(parametersCurrent);
    return new TwitterIteratorProxy<ITwitterResult<ITweetDTO[]>, ITweet, number>(pageIterator, twitterResult =>
      this._client.factories.createTweets(twitterResult?.model));
  }

  public async getRetweetsOfMeTimelineAsync(parameters?: IGetRetweetsOfMeTimelineParameters): Promise<ITweet[]> {
    let parametersCurrent: IGetRetweetsOfMeTimelineParameters;
    if (parameters) {
      parametersCurrent = parameters;
    } else {
      parametersCurrent = new GetRetweetsOfMeTimelineParameters();
    }

    let iterator = this.getRetweetsOfMeTimelineIterator(parametersCurrent);
    return [...(await iterator.nextPageAsync())];
  }

  public getRetweetsOfMeTimelineIterator(parameters?: IGetRetweetsOfMeTimelineParameters): ITwitterIterator<ITweet, number> {   // long?
    let parametersCurrent: IGetRetweetsOfMeTimelineParameters;
    if (parameters) {
      parametersCurrent = parameters;
    } else {
      parametersCurrent = new GetRetweetsOfMeTimelineParameters();
    }

    let pageIterator = this._timelinesRequester.getRetweetsOfMeTimelineIterator(parametersCurrent);

    return new TwitterIteratorProxy<ITwitterResult<ITweetDTO[]>, ITweet, number>(pageIterator,  // long?
      twitterResult => this._client.factories.createTweets(twitterResult?.model));
  }

  private isIUserIdentifier(userIdOrUsernameOrUserIdentifierOrParameters: any): userIdOrUsernameOrUserIdentifierOrParameters is IUserIdentifier {
    return (userIdOrUsernameOrUserIdentifierOrParameters as IUserIdentifier).screenName !== undefined;
  }
}
